mlcomp/parallelm/mlpiper/wizard_flow.py

```diff
@@ -0,0 +1,137 @@
+
+from pypsi import wizard as wiz
+from parallelm.mlpiper.wizard_shell import CommandName, EXIT_CODE
+
+
+# TODO:
+# make connection type an options list
+
+# remove header, customize?
+# steps dependency
+# ask for dir and generate __init__ and main.py with class
+
+# validate file is component json
+# if quit - exit totally
+# Discuss: are these required: label, description
+
+
+class State(object):
+    class YesNoWizard(object):
+        def __init__(self, name="yesno", header=""):
+            self.wizard = wiz.PromptWizard(
+                name=name,
+                description="Provide values for component connection attributes",
+                steps=[
+                    wiz.WizardStep(id="answer", name=header, help="Answer Yes or No",
+                                   default="Yes", validators=wiz.boolean_validator)
+                ]
+            )
+
+        def run(self, shell):
+            return self.wizard.run(shell, print_header=False)
+
+    def __init__(self):
+        pass
+
+    def _yes_no_wizard(self, shell, header):
+        w = self.YesNoWizard(name="yesno", header=header)
+        ns = w.run(shell)
+        if ns:
+            return ns.answer
+        else:
+            return False
+
+    def get_input(self, shell):
+        raise Exception("Not implemented")
+
+    def run(self, shell, step_input):
+        raise Exception("Not implemented")
+
+
+class StartState(State):
+    def get_input(self, shell):
+        return True
+
+    def run(self, shell, step_input):
+        return AddComponentState()
+
+
+class FinishState(State):
+    pass
+
+
+class AddComponentState(State):
+
+    def get_input(self, shell):
+        return True
+
+    def run(self, shell, step_input):
+        ret_code = shell.execute(CommandName.COMPONENT)
+        if ret_code == EXIT_CODE:
+            return FinishState()
+        return AddInputState()
+
+
+class AddInputState(State):
+    _header = "Do you want to add an input connection?"
+
+    def get_input(self, shell):
+        return self._yes_no_wizard(shell, AddInputState._header)
+
+    def run(self, shell, step_input):
+        if step_input:
+            shell.execute(CommandName.INPUT)
+            AddInputState._header = "Do you want to add an additional input connection?"
```

Comment:
It is on the class level because I want to have one message for the first run of this step and another message starting with the second step:
"Do you want to add an input connection?"
and
"Do you want to add an additional input connection?"

When it is on class level both messages are defined in the same class, in the same code block. If it will be an argument, I'll have to pass it in different code blocks which can be confusing.
I will add a comment and we take a look one more time.

Reply:
The beauty of state machine is its simplicity. In general, a given state data should not be changed over time, otherwise you'll end up with complex and less clear code. In you case, I would expect to have two states - one for the first question and the other for the second one. Moreover, you create states on the fly. In a pure state machine all the states should be defined and allocated before you run the state machine. Maybe it's more code, but believe me it's much simpler. And BTW, you can create states with common code by inheriting from a common patent (.e.g AddInputActionStateBase)

Reply:
fixed, please take a look